tests/xdc/nexus_state_replication_test.go

```diff
@@ -214,6 +234,87 @@ func (s *NexusStateReplicationSuite) TestNexusOperationEventsReplicated() {
 	s.NoError(run.Get(ctx, nil))
 }
 
+// TestNexusCallbackReplicated tests that nexus callback and state updates are replicated
+// across clusters and callback can work as expected when failover happens.
+// General outline:
+// 1. Start two clusters, cluster1 set to active, cluster2 set to standby.
+// 2. Start a workflow will callback on cluster1.
+// 3. Terminate the workflow to trigger the callback. An error is injected to always fail the callback.
+// 4. Check the callback state changes are replicated to cluster2.
+// 5. Failover to cluster2 and unblock the callback by removing the injected error.
+// 6. Wait for the callback to complete on cluster2.
+func (s *NexusStateReplicationSuite) TestNexusCallbackReplicated() {
+	failCallback := atomic.NewBool(true)
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if failCallback.Load() {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, "Injected error to trigger callback retry")
+			return
+		}
+		fmt.Fprintln(w, "Callback succeeded")
+	}))
+	defer ts.Close()
+
+	ctx := tests.NewContext()
+	ns := s.createGlobalNamespace()
+
+	sdkClient1, err := sdkclient.Dial(sdkclient.Options{
+		HostPort:  s.cluster1.GetHost().FrontendGRPCAddress(),
+		Namespace: ns,
+	})
+	s.NoError(err)
+	sdkClient2, err := sdkclient.Dial(sdkclient.Options{
+		HostPort:  s.cluster2.GetHost().FrontendGRPCAddress(),
+		Namespace: ns,
+	})
+	s.NoError(err)
+
+	tv := testvars.New(s.T())
+	startResp, err := sdkClient1.WorkflowService().StartWorkflowExecution(ctx, &workflowservice.StartWorkflowExecutionRequest{
+		Namespace:    ns,
+		WorkflowId:   tv.WorkflowID(),
+		WorkflowType: tv.WorkflowType(),
+		TaskQueue:    tv.TaskQueue(),
+		RequestId:    uuid.New(),
+		CompletionCallbacks: []*commonpb.Callback{
+			{
+				Variant: &commonpb.Callback_Nexus_{
+					Nexus: &commonpb.Callback_Nexus{
+						Url: ts.URL,
+					},
+				},
+			},
+		},
+	})
+	s.NoError(err)
+
+	// Terminate the workflow to trigger the callback.
+	err = sdkClient1.TerminateWorkflow(ctx, tv.WorkflowID(), startResp.RunId, "terminate workflow to trigger callback")
+	s.NoError(err)
+
+	// Check callback state changes are replicated to cluster2.
+	s.waitCallback(ctx, sdkClient2, &commonpb.WorkflowExecution{
+		WorkflowId: tv.WorkflowID(),
+		RunId:      startResp.GetRunId(),
+	}, func(callback *workflow.CallbackInfo) bool {
+		return callback.Attempt > 2
+	})
+
+	// Failover to cluster2.
+	s.failover(ns, s.clusterNames[1], 2, s.cluster1.GetFrontendClient())
+
+	// Unblock callback after failover.
+	failCallback.Store(false)
+
+	// Check callback can complete on cluster2 after failover.
+	s.waitCallback(ctx, sdkClient2, &commonpb.WorkflowExecution{
```

Comment:
It would be nice to check this on cluster 1 just in case.

Reply:
Updated the logic to check state on both clusters.